Optimize multibuy database queries

patrickreiffenstein opened the pull request:
Fixes #357.

patrickreiffenstein committed:
Optimized multibuy hint

patrickreiffenstein committed:
Merge branch 'next' into fixIssue357

patrickreiffenstein committed:
Removed duplicate queries

patrickreiffenstein committed:
Removed debugging I left behind

krestenlaust commented on stregsystem/models.py:
```diff
@@ -128,15 +128,15 @@ def execute(self):
         self.member.fulfill(transaction)
 
+        # Collect all the sales of the order
+        Sales = []
```
```suggestion
        sales = []
```

krestenlaust commented on stregsystem/models.py:
```diff
@@ -128,15 +128,15 @@ def execute(self):
             for i in range(item.count):
                 s = Sale(member=self.member, product=item.product, room=self.room, price=item.product.price)
-                s.save()
+                Sales.append(s)
```
```suggestion
                sales.append(s)
```

krestenlaust reviewed (commented):
I'll have to run it locally to check things over, and verify that we have tests in place

krestenlaust commented:
I'll try to review this before summer ends

JakobTopholt commented:
@Mast3rwaf1z Vil du ikke være en king og review den her så Kresten kan få lidt fritid?

krestenlaust commented:
♥️ jeg vil dog stadig gerne merge den

Mast3rwaf1z commented:
> @Mast3rwaf1z Vil du ikke være en king og review den her så Kresten kan få lidt fritid?

Tager lige et kig senere, var lige til forelæsning

Mast3rwaf1z commented on stregsystem/models.py:
```diff
@@ -128,15 +128,15 @@ def execute(self):
-                s.save()
+                Sales.append(s)
+        # Save all the sales
+        Sale.objects.bulk_create(Sales)
```
```suggestion
        Sale.objects.bulk_create(sales)
```

Mast3rwaf1z commented on stregsystem/views.py:
```diff
@@ -715,14 +713,23 @@ def api_quicksale(request, room, member: Member, bought_ids):
+        # Get the amount of unique items bought
+        unique_product_dict = {}
+        for unique_id in bought_ids:
+            if str(unique_id) not in unique_product_dict:
+                unique_product_dict[str(unique_id)] = 1
+            else:
+                unique_product_dict[str(unique_id)] += 1
```
```suggestion
            if unique_id not in unique_product_dict:
                unique_product_dict[unique_id] = 1
            else:
                unique_product_dict[unique_id] += 1
```
what's the point in converting from an int object to string all the time?

Mast3rwaf1z replied:
you can also optimize it further by just using dict comprehension:

```py
unique_product_dict = {i:bought_ids.count(i) for i in bought_ids}
```

Mast3rwaf1z commented on stregsystem/views.py:
```diff
@@ -715,14 +713,23 @@ def api_quicksale(request, room, member: Member, bought_ids):
+                unique_product_dict[str(unique_id)] += 1
+
+        # Add the given amount of different products
+        for i in unique_product_dict:
             product = Product.objects.get(
                 Q(pk=i),
                 Q(active=True),
                 Q(deactivate_date__gte=time_now) | Q(deactivate_date__isnull=True),
                 Q(rooms__id=room.id) | Q(rooms=None),
```
```suggestion
        for key, value in unique_product_dict.items():
            product = Product.objects.get(
                Q(pk=key),
                Q(active=True),
                Q(deactivate_date__gte=time_now) | Q(deactivate_date__isnull=True),
                Q(rooms__id=room.id) | Q(rooms=None),
```

Mast3rwaf1z commented on stregsystem/views.py:
```diff
@@ -715,14 +713,23 @@ def api_quicksale(request, room, member: Member, bought_ids):
                 Q(rooms__id=room.id) | Q(rooms=None),
             )
-            products.append(product)
+            products.extend([product for x in range(unique_product_dict[str(i)])])
```
```suggestion
            products.extend([product for _ in range(value)])
```

Mast3rwaf1z reviewed (changes requested):
Some small things, i think improves readability of the code, and removed some redundant constructor calls?

also i believe by using `_` in a for loop prevents memory being allocated to the variable. really insignificant but hey, its good to have imo

patrickreiffenstein committed:
Apply suggestions from code review

Added the suggestions from Kresten and Thomas, which seems to be some good changes

Co-authored-by: Kresten Laust <krestenlaust@gmail.com>
Co-authored-by: Thomas Jensen <32744171+Mast3rwaf1z@users.noreply.github.com>

patrickreiffenstein committed:
Merge branch 'next' into fixIssue357

Mast3rwaf1z commented on stregsystem/views.py:
```diff
@@ -147,15 +147,14 @@ def _multibuy_hint(now, member):
+    recent_unique_purchases = recent_purchases.values('product').distinct().annotate(total=Count('product'))
 
     # add hint for multibuy
     if number_of_recent_distinct_purchases > 1:
         sale_dict = {}
-        for sale in recent_purchases:
-            if not str(sale.product.id) in sale_dict:
-                sale_dict[str(sale.product.id)] = 1
-            else:
-                sale_dict[str(sale.product.id)] = sale_dict[str(sale.product.id)] + 1
+        for unique_sale in recent_unique_purchases:
+            sale_dict[str(unique_sale['product'])] = unique_sale['total']
+
```
inconsistent naming, sale and purchase

patrickreiffenstein replied:
Fixed

Mast3rwaf1z reviewed (approved):
looks good, minor annoying naming issue

patrickreiffenstein committed:
Fix inconsistent naming

Mast3rwaf1z commented on stregsystem/views.py:
```diff
@@ -841,14 +840,23 @@ def api_quicksale(request, room, member: Member, bought_ids):
-            products.append(product)
+            products.extend([product for _ in range(value)])
     except Product.DoesNotExist:
         return "Invalid product id", 400, i
```
```suggestion
        return "Invalid product id", 400, key
```

patrickreiffenstein replied:
Should be fixed now

patrickreiffenstein committed:
Fix missed variable renaming in test

patrickreiffenstein committed:
fixup! Fix missed variable renaming in test

patrickreiffenstein committed:
Merge branch 'next' into fixIssue357

patrickreiffenstein committed:
Merge branch 'f-klubben:next' into fixIssue357

patrickreiffenstein committed:
Merge branch 'f-klubben:next' into fixIssue357

patrickreiffenstein commented:
My fork is broken and I don't know why it acts that way, so these changes are moved to #589.
